Source/Core/VideoCommon/PixelShaderGen.cpp

```diff
@@ -732,8 +752,135 @@ uint WrapCoord(int coord, uint wrap, int size) {{
   }
 }
 
+void WriteCustomShaderStructImpl(ShaderCode* out, u32 num_stages, bool per_pixel_lighting,
+                                 const pixel_shader_uid_data* uid_data)
+{
+  out->Write("\tCustomShaderData custom_data;\n");
+
+  if (per_pixel_lighting)
+  {
+    out->Write("\tcustom_data.position = WorldPos;\n");
+    out->Write("\tcustom_data.normal = Normal;\n");
+  }
+  else
+  {
+    out->Write("\tcustom_data.position = float3(0, 0, 0);\n");
+    out->Write("\tcustom_data.normal = float3(0, 0, 0);\n");
+  }
+
+  if (uid_data->genMode_numtexgens == 0) [[unlikely]]
+  {
+    out->Write("\tcustom_data.texcoord[0] = float3(0, 0, 0);\n");
+  }
+  else
+  {
+    for (u32 i = 0; i < uid_data->genMode_numtexgens; ++i)
+    {
+      out->Write("\tif (tex{0}.z == 0.0)\n", i);
+      out->Write("\t{{\n");
+      out->Write("\t\tcustom_data.texcoord[{0}] = tex{0};\n", i);
+      out->Write("\t}}\n");
+      out->Write("\telse {{\n");
+      out->Write("\t\tcustom_data.texcoord[{0}] = float3(tex{0}.xy / tex{0}.z, 0);\n", i);
+      out->Write("\t}}\n");
+    }
+  }
+
+  for (u32 i = 0; i < uid_data->genMode_numindstages; ++i)
+  {
+    if ((uid_data->nIndirectStagesUsed & (1U << i)) != 0)
+    {
+      u32 texcoord = uid_data->GetTevindirefCoord(i);
+
+      // Quirk: when the tex coord is not less than the number of tex gens (i.e. the tex coord
+      // does not exist), then tex coord 0 is used (though sometimes glitchy effects happen on
+      // console). This affects the Mario portrait in Luigi's Mansion, where the developers forgot
+      // to set the number of tex gens to 2 (bug 11462).
+      if (texcoord >= uid_data->genMode_numtexgens)
+        texcoord = 0;
+
+      out->Write("\t{{\n");
+      out->Write("\t\tint2 fixpoint_uv = int2(custom_data.texcoord[{}].xy", texcoord);
+      out->Write(" * float2(" I_TEXDIMS "[{}].zw * 128));\n", texcoord);
+      out->Write("\t\tint2 tempcoord = fixpoint_uv >> " I_INDTEXSCALE "[{}].{};\n", i / 2,
+                 (i & 1) ? "zw" : "xy");
+      out->Write("\t\tcustom_data.texcoord[{0}] = float3(tempcoord / float2(" I_TEXDIMS
+                 "[{0}].zw * 128), 0);\n",
+                 texcoord);
+      out->Write("\t}}\n");
+    }
+  }
```

Comment:
Yeah.  Essentially the goal is to get the texture samplers and the texture coordinates used by the game for a given draw call.  At the moment draw-call = texture, which of course isn't really right (ex: multiple textures defined for a draw) but it's hard for users to point at a draw call without some sort of editor.

The goal of this block was to grab the texture coordinates, regardless of whether it is an indirect or direct call.  This normalizes the data, so the user doesn't have to care (they probably don't).

> I'm pretty sure you can use the same texture coordinate for multiple indirect stages (e.g. both of them use texture coordinate 0 for different textures)

Yes, this is fine.  Just want the user to say for this texture, give me the sampler and texture coordinates.

> Maybe it'd be better to just get rid of this entirely for now.

I'm fine doing that but still not clear what the long-term solution would be?  If the draw-call = texture is the issue, I can certainly fix that in the future.  I vaguely recall before I had this, lots of textures just weren't mapping right.

Reply:
(Will address thus more fully later - I don't entirely understand what you mean by draw-call = texture, and will need to check more to understand better).

> Yes, this is fine. Just want the user to say for this texture, give me the sampler and texture coordinates.

It's fine in the case of https://github.com/dolphin-emu/dolphin/pull/11300#discussion_r1292906918, but not here where the same texture coordinate will get scaled multiple times.

Reply:
> The goal of this block was to grab the texture coordinates, regardless of whether it is an indirect or direct call. This normalizes the data, so the user doesn't have to care (they probably don't).

<details><summary>Explanation attempt 1</summary>

This isn't something that makes sense to do. Based on [libogc's `GX_SetIndTexCoordScale`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L3252-L3264), the main use of the indirect scale is so that you can use the same texture coordinate for both a regular texture and an indirect texture, but have different sizes for the two.

For each texture coordinate, `bpmem.texcoords[i].s.scale_minus_1` gives a scale on the horizontal axis, and `t` on the vertical axis, which are _generally_ set to the size of the corresponding texture in texels ([libogc's `GX_SetTexCoordScaleManually`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574) says that this happens automatically in most cases, but can be overridden if needed). So for a 256 by 256 texture, you would supply input coordinates between (0, 0) and (1, 1), and then these would be scaled to (0, 0) to (256, 256) when actually sampling the texture. GLSL's `texture` function doesn't expect this multiplication by the texture size, so we have to divide it by the texture's actual size ahead of time; the texture's actual size is stored in `I_TEXDIMS[i].xy` while this scale is `I_TEXDIMS[i].zw`. (Manual texture sampling uses `texelFetch` which does require multiplication by the texture size, and also manual wrapping; there's an extra complication here about the texture size the game uses being stored in `I_TEXDIMS[i].xy`, but custom textures possibly having a _different_ size that needs to be used instead, which involves querying the size in GLSL.)
</details>

... OK, that's a bit of a confusing explanation. I'm going to try it in a different way:

Games usually specify textures coordinates as UV coordinates where (0, 0) is the top-left of the texture and (1, 1) is the bottom-right of the texture. This includes the GameCube/Wii; Modern graphics APIs have functions such as `texture` where you can just provide UVs and everything works. This is probably the coordinate system you want to expose for shaders.

But for the GameCube/Wii, sampling textures is done with ST coordinates. For instance, for a 256 by 256 texture, (0, 0) is the top-left and (255, 255) is the bottom-right (... ish, there may be an off-by-one here). I've seen this UV versus ST distinction in graphics textbooks, but Dolphin doesn't really use it currently. There's a scale factor that UV coordinates get multiplied by to convert to ST coordinates. That said, this is _mostly_ hidden from the GX API: unless [`GX_SetTexCoordScaleManually`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574) is used, the scale (in `bpmem.texcoords[texcoord].s.scale_minus_1` for `s`, and a similar value for `t`, and in shaders `I_TEXDIMS[texcoord].zw`) will match the one for the texture being used (in `AllTexUnits.GetUnit(texmap).texImage0` or `I_TEXDIMS[texmap].xy` - note the difference in indexing here).

This does pose an issue when using the same texture coordinate for multiple textures, if those textures are of different sizes (for instance one is 256 by 256 and the second is 128 by 128); the scale will only match one of those. If the scale factor is 256, then the 128 by 128 texture will be drawn at half-scale horizontally (as in it'll be drawn twice for every time the bigger texture is drawn, or rather 4 times since this applies on both axes). I don't think there's any mechanism to deal with this for most textures that end up getting drawn to the screen, but there _is_ a mechanism for doing this for indirect textures: [`GX_SetTexCoordScaleManually`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574), which modifies `bpmem.texscale[indstage/2]`. This lets you scale down the texture coordinates when sampling a texture for an indirect texture operation (i.e. when you're reading a texture so that you can then add the values read from that texture to another texture coordinate, possibly transformed by a matrix). So, if you have a 256 by 256 main texture, and also a 128 by 128 texture you want to use as an indirect texture, you can set `bpmem.texscale[0].ss0` and `ts0` to 1, and thus the ST coordinates will be at appropriate scales for both textures.

The key issue here is that this kind of scaling is meaningless for UV coordinates, where the texture's size isn't a factor at all. I'd expect most custom shaders to use `texture` (and not e.g. `texelFetch`), so applying the scaling wouldn't make sense. Furthermore, even if ST coordinates were used, it would be more useful to provide ST coordinates on the main texture's scale, since that's what ends up on the screen eventually. (There's also a secondary issue of multiple indirect stages where all of the stages use the same texture coord; if you have two 128 by 128 indirect textures, the current code would divide s and t by 4, instead of 2.)

This is a lot of text, but hopefully it explains what the point of all of these are well enough.

Reply:
I just want to say thank you so much for taking the time to write this in-depth explanation.  I've read it a couple times and your reasoning makes a lot of sense now.  I had heard of ST coordinates but always assumed they were the same as UV coordinates.  Maybe this distinction is special to the Cube/Wii?

Regardless, I see now why I don't need the indirect texture uv transformation and only need to capture those texture map ids.

Thanks again for the explanation.  I have it all sorted now I think!

Reply:
I tried to dig up more information regarding ST versus UV and can't find anything concrete about the differences between them (either online or in various GC patents or in my graphics textbook), just that both are used by convention for textures. It would be convenient if there was a commonly used set of terminology for the difference between ranging between 0 and 1 versus 0 and width, but it doesn't seem like there actually is one. This doesn't change the actual GameCube/Wii behavior though (which should be as I described above); I'm just clarifying in case you see the terms elsewhere.

Reply:
The original difference was that UV were surface coordinates (the range is whatever you want), and ST are texture coordinates (texture matrices applied, wrap mode is applied to fit within 0-1, sometimes the coordinates are unnormalized), but this hasn't been true in practice for many many years as hardware advancements happened and as shaders took off.

UV is now the standard, ST died out as it isn't a meaningful unit unless you make GPUs.

Reply:
> This does pose an issue when using the same texture coordinate for multiple textures, if those textures are of different sizes (for instance one is 256 by 256 and the second is 128 by 128); the scale will only match one of those. If the scale factor is 256, then the 128 by 128 texture will be drawn at half-scale horizontally (as in it'll be drawn twice for every time the bigger texture is drawn, or rather 4 times since this applies on both axes).

The standard way to handle this is to just eat up multiple texgens from the same texture coordinate.

As mentioned, this is secretly handled for you with the GX API -- right before a display list or a draw begins, it sets the texcoord scale registers to be the size of the texture. The GX's per-pixel pipeline is *all* unnormalized, the per-vertex XF unit generates unnormalized coordinates as one of the last things it does.

This means that indirect texture offsets are unnormalized, too. If I read +128 from an indirect texture, that's going to move +128 pixels in the final unnormalized coordinate no matter the original size. But the IndTexMtx can be used to adjust for this.

> I don't think there's any mechanism to deal with this for most textures that end up getting drawn to the screen, but there is a mechanism for doing this for indirect textures: [GX_SetTexCoordScaleManually](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574), which modifies bpmem.texscale[indstage/2]

I think you meant to link GXSetIndTexCoordScale here?

Reply:
Thanks for the info on UV versus ST.

It also sounds like the distinction I wanted was normalized versus unnormalized, where normalized is for ranges of 0 to 1 while unnormalized is for ranges of 0 to width/height. (... or at least can be, I guess, and you'd also have to consider wrapping at some point, but eh.)

> I think you meant to link GXSetIndTexCoordScale here?

Yes, I meant to copy the link to [GX_SetIndTexCoordScale](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574) from my first abandoned explanation, but I must not have been paying attention and copied the wrong one.

> This means that indirect texture offsets are unnormalized, too. If I read +128 from an indirect texture, that's going to move +128 pixels in the final unnormalized coordinate no matter the original size. But the IndTexMtx can be used to adjust for this.

There's also an [`indtexbias`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L802-L817) parameter to [`GX_SetTevIndirect`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L3206-L3216) (which ends up in `bpmem.tevind[stage].bias`), which will subtract 128 from it in most cases, prior to the matrix being applied. (The exception is when the indirect format is not `ITF_8`, which is used in the Skyward Sword map (see #9876 and [my writeup](https://gist.github.com/Pokechu22/dfebe2758d23faeba48bb7c9fb8557a2).)

> The standard way to handle this is to just eat up multiple texgens from the same texture coordinate.

Yeah, that's what I've generally seen. If a game does this, then there's no special handling needed for graphics mods; an array listing the texture coordinate corresponding to each texture will work fine. (PixelShaderGen.cpp gets those generated texture coordinates as an input, so even if the model itself has only one texture coordinate or if the texture coordinates are generated from positions, at this point everything will be fine.)

> The GX's per-pixel pipeline is all unnormalized, the per-vertex XF unit generates unnormalized coordinates as one of the last things it does.

It's worth noting that Dolphin does this multiplication by texture size in the pixel shader. This is probably suboptimal since the texture size won't change, but it does mean that (apart from when per-pixel lighting is enabled) everything in the vertex shader is based on `xfmem` and everything in the pixel shader is based on `bpmem`.

[Yagcd](http://hitmen.c02.at/files/yagcd/yagcd/chap5.html#sec5.11.1) says that the relevant variables are `SU_SSIZE0` and `SU_TSIZE0`, which I think refers to the setup unit which is between the transform unit and the TEV (I think it's responsible for making triangles, although the rasterizer is also listed separately).

Handling both the multiplication by the scale in the vertex shader would be nice, but we'd still need to divide by the texture size for GLSL's `texture` and similar to work properly, both for Dolphin's use and for graphics mods. (And _that_ division couldn't be done in the vertex shader if we want fixed-point math and indirect textures to work).